<!--
% \VignetteIndexEntry{How to write recipes for new resources for the AnnotationHub}
% \VignetteDepends{AnnotationHub}
% \VignetteEngine{knitr::knitr}
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

# AnnotationHub: How to write recipes for new resources for the AnnotationHub

**Package**: `r Biocpkg("AnnotationHub")`<br />
**Authors**: `r packageDescription("AnnotationHub")[["Author"]] `<br />
**Modified**: 11 October, 2014<br />
**Compiled**: `r date()`


## Overview of the process

If you are reading this it is (hopefully) because you intend to write
some code that will allow the processing of online resources into R
objects that are to be made available via that the
AnnotationHub package.  In order to do this you will have
to do three basic steps (outlined below).  These steps will have you
writing two functions and then calling a third function to do some
automatic set up for you.  The 1st function will contain instructions
on how to process data that is stored online into metadata for
describing your new R resources for the AnnotationHub.  And the 2nd
function is for describing how to take these online resources and
transform them into an R object that is useful to end users.

## Introducing AnnotationHubMetadata Objects

The AnnotationHubData package is a complementary package to
the AnnotationHub package that provides a place where we
can store code that processes online resources into R objects suitable
for access through the AnnotationHub package.  But before
you can understand the requirements for this package it is important
that you 1st learn about the objects that are used as
intermediaries between the hub and its web based repository behind the
scenes.  That means that you need to know about
AnnotationHubMetadata objects.  These objects store the
metadata that describes an online resource.  And if you want to see a
set of online resources added to the repository and maintained, then
it will be necessary to become familiar with the
AnnotationHubMetadata constructor.  For each online
resource that you want to process into the AnnotationHub, you will
have to be able to construct an AnnotationHubMetadata object that describes it in detail and that specifies where the recipe
function lives.



## Step 1: Writing your AnnotationHubMetadata generating function
  
The 1st function you need to provide is one that processes some online
resources into AnnotationHubMetadata objects.  This function
MUST return a list of AnnotationHubMetadata objects.  It can
rely on other helper functions that you define, but ultimately it (and it's helpers) need to contain all of the instructions needed to find resources and process those resources into AnnotationHubMetadata objects.

The following example function takes files from the latest release of inparanoid and processes them into AnnotationHubMetadata objects using Map.
The calling of the Map function is really the important part of this
function, as it shows the function creating a series of
AnnotationHubMetadata objects.  Prior to that, the function
was just calling out to other helper functions in order to process the
metadata so that it could be passed to the
AnnotationHubMetadata constructor using Map.
Notice how one of the fields specified by this function is the Recipe,
which indicates both the name and location of the recipe function.  We
expect most people will want to submit their recipe to the same
package as they are submitting their metadata processing function.

```{r, exampleInpProcessing}
makeinparanoid8ToAHMs <- function(currentMetadata){
    baseUrl <- 'http://inparanoid.sbc.su.se/download/current/Orthologs_other_formats'
    ## Make list of metadata in a helper function
    meta <- .inparanoidMetadataFromUrl(baseUrl)
    ## then make AnnotationHubMetadata objects.
    Map(AnnotationHubMetadata,
        Description=meta$description,
        Genome=meta$genome,
        SourceFile=meta$sourceFile, 
        SourceUrl=meta$sourceUrl,
        SourceVersion=meta$sourceVersion,
        Species=meta$species,
        TaxonomyId=meta$taxonomyId,
        Title=meta$title,
        RDataPath=meta$rDataPath,
        MoreArgs=list(
          Coordinate_1_based = TRUE,
          DataProvider = baseUrl,
          Maintainer = "Marc Carlson <mcarlson@fhcrc.org>",
          RDataClass = "SQLiteFile",
          RDataDateAdded = Sys.time(),
          RDataVersion = "0.0.1",
          Recipe = c("inparanoid8ToDbsRecipe", 
                     package="AnnotationHubData"),
          Tags = c("Inparanoid", "Gene", "Homology", "Annotation")))
}
```


## Step 2: Writing your recipe 

The 2nd kind of function you need to write is called a recipe
function.  It always must take an single argument that must be an AnnotationHubMetadata object.  The job of a recipe function is to use the metadata from an AnnotationHubMetadata object to produce an R object or data file that will be retrievable from the AnnotationHub service later on.  Below is a recipe function that calls some helper functions to generate an inparanoid database object from the metadata stored in it's AnnotationHubMetadata object.

```{r, exampleRecipe}
inparanoid8ToDbsRecipe <- function(ahm){
    require(AnnotationForge)
    inputFiles <- metadata(ahm)$SourceFile
    dbname <- makeInpDb(dir=file.path(inputFiles,""),
                        dataDir=tempdir())
    db <- loadDb(file=dbname)
    outputPath <- file.path(metadata(ahm)$AnnotationHubRoot,
                            metadata(ahm)$RDataPath)
    saveDb(db, file=outputPath) 
    outputFile(ahm)
}
```


## Step 3: Calling the makeAnnotationHubResource helper

Finally you will need to call the
makeAnnotationHubResource function to do some setup.  This
function only has two required arguments.  The 1st is basically the
name of a class that describes the kind of resource you are writing
code to import.  It just needs to be a unique name and will be used internally to create a class for dispatch.  The 2nd argument
is the name of your metadata processing function from step one.  Once
you have finished this, the only step left is to export the class name
into the NAMESPACE (remember that this is that string you are providing as your 1st argument), and then add this code to the AnnotationHubData repository.  We are planning to set up a bridge to github so that you can give us a pull request.  

```{r, makeAnnotationHubResource, eval=FALSE}
makeAnnotationHubResource("Inparanoid8ImportPreparer",
                          makeinparanoid8ToAHMs)
```


## Step 4: Testing your new recipe

At this point you are probably wondering if this recipe will even work.  Well there is a way to test it.  You can call the function that will try to update the resources for your recipe.  That function is called updateResources(). The updateResources() function takes several arguments.  These are:

* ahroot: represents the place where you would like any generated files go.
* BiocVersion: character vector for versions of Bioc your objects will be compatible with
* preparerClasses: vector of which classes you want to process recipes for.
* insert: try to insert metadata into the DB? You don't get to do this.  So it should be FALSE when you call this or you will get an error.
* metadataOnly: should the recipes be run to create the processed files?

```{r, updateResources, eval=FALSE}
library(AnnotationHubData)
ahroot <- getwd()
BiocVersion <- c("3.0")  

mdinp = updateResources(ahroot, 
                        BiocVersion,
                        preparerClasses = "Inparanoid8ImportPreparer",
                        insert = FALSE,
                        metadataOnly=TRUE)     
```

When you call updateResources() like this, a set of AnnotationHubMetadata objects should be returned.  You should inspect those to make sure that they contain all the metadata that you intended to use.  And if you run it again with metadataOnly set to FALSE, then it should generate the final objects for you in the ahroot directory that you specified.  If both of these things happen to your satisfaction then you can contact us about your new recipe.


## Session Information

```{r, SessionInfo, echo=FALSE}
sessionInfo()
```





